### ISSUE TITLE
[CT-284] [Bug] Unable to run dbt inside a docker container due to missing dependencies

### ISSUE THREAD

**Issue by alexrosenfeld10** [dbt-labs/dbt-core#4784]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Running `dbt deps` inside a Dockerfile has no effect on the final resulting image. This is pretty much a full blocker for me, and I feel like I'm going crazy that this isn't already an issue anyone has raised. I must be missing something obvious, right?

### Expected Behavior

Running `dbt deps` should persist the generated `dbt_packages` in the final resulting image. Immutable deploys are super important so you know exactly what you're getting every time you ship code. This means the deps are bundled in the image at build time. That way if anything is down (like the servers where the dbt dependencies are hosted), my image still has everything it needs inside it to start up. Or, if I need to ship the _exact same image_ again, I can.

### Steps To Reproduce

Set up this docker file:

```Dockerfile
FROM ghcr.io/dbt-labs/dbt-core:1.0.latest
RUN pip install dbt-snowflake
COPY / ./
RUN dbt deps
ENTRYPOINT ["dbt", "--log-format", "json", "--warn-error", "run"]
```

The resulting image will not contain the `dbt_packages` directory, even though the `dbt deps` command was run. For example:

```shell
❯ docker run -it --entrypoint=/bin/sh my-image:my-tag
# ls
dbt_project.yml  models	     seeds	tests
analyses      macros	       packages.yml  snapshots
#
```

There is no `dbt_packages` directory. 

### Relevant log output

```shell
17:09:24  + docker build -t <my-company-jfrog-url>:2022-02-24.43-83023aa-rel -f Dockerfile .
17:09:24  Sending build context to Docker daemon  600.1kB

17:09:24  Step 1/5 : FROM ghcr.io/dbt-labs/dbt-core:1.0.latest
17:09:24   ---> afbbb0909dea
17:09:24  Step 2/5 : RUN pip install dbt-snowflake
17:09:24   ---> Using cache
17:09:24   ---> 218706b2f74e
17:09:24  Step 3/5 : COPY / ./
17:09:25   ---> 614cce69e3d8
17:09:25  Step 4/5 : RUN dbt deps
17:09:25   ---> Running in 2b9fc7ee9938
17:09:27  22:09:27  Running with dbt=1.0.2
17:09:28  22:09:28  Installing dbt-labs/dbt_utils
17:09:29  22:09:28    Installed from version 0.8.1
17:09:29  22:09:28    Up to date!
17:09:29  22:09:28  Installing calogica/dbt_expectations
17:09:29  22:09:29    Installed from version 0.5.1
17:09:29  22:09:29    Updated version available: 0.5.2
17:09:29  22:09:29  Installing calogica/dbt_date
17:09:29  22:09:29    Installed from version 0.5.3
17:09:29  22:09:29    Up to date!
17:09:29  22:09:29  
17:09:29  22:09:29  Updates available for packages: ['calogica/dbt_expectations']                 
17:09:29  Update your versions in packages.yml, then run dbt deps
17:09:30  Removing intermediate container 2b9fc7ee9938
17:09:30   ---> 1420e054b93a
17:09:30  Step 5/5 : ENTRYPOINT ["dbt", "--log-format", "json", "--warn-error", "run"]
17:09:30   ---> Running in 9a6e49e3fe2a
17:09:31  Removing intermediate container 9a6e49e3fe2a
17:09:31   ---> aa6f5ca2bcba
17:09:31  Successfully built aa6f5ca2bcba
17:09:31  Successfully tagged <my-company-jfrog-url>:2022-02-24.43-83023aa-rel
```
```


### Environment

```markdown
- OS:
- Python:
- dbt:
```

### Additional Context

other folks are having the same issue: https://getdbt.slack.com/archives/C2JRRQDTL/p1645743857126709

**Comment by alexrosenfeld10** [comment]
OK, I think this is because of this line: https://github.com/dbt-labs/dbt-core/blob/main/docker/Dockerfile#L53

it makes the default directory immutable at build time, which is super not-obvious for the user.

Here's how I'm getting around it:

```Dockerfile
FROM ghcr.io/dbt-labs/dbt-snowflake:1.0.latest

RUN dbt --version

WORKDIR /usr/my-project
COPY . .
RUN dbt deps

# TODO obviously this kind of thing needs to be parameterized in future.
ENTRYPOINT ["dbt", "--log-format", "json", "--warn-error", "run"]
```

**Comment by iknox-fa** [comment]
Hmm.  It seems to be working correctly for me:
```
> docker build -t deps_test .     
[+] Building 4.2s (9/9) FINISHED                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                               0.0s
 => => transferring dockerfile: 37B                                                                                                                0.0s
 => [internal] load .dockerignore                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                    0.0s
 => [internal] load metadata for ghcr.io/dbt-labs/dbt-core:1.0.latest                                                                              0.0s
 => [1/4] FROM ghcr.io/dbt-labs/dbt-core:1.0.latest                                                                                                0.0s
 => [internal] load build context                                                                                                                  0.1s
 => => transferring context: 1.29MB                                                                                                                0.1s
 => CACHED [2/4] RUN pip install dbt-snowflake                                                                                                     0.0s
 => [3/4] COPY / ./                                                                                                                                0.0s
 => [4/4] RUN dbt deps                                                                                                                             3.1s
 => exporting to image                                                                                                                             0.9s 
 => => exporting layers                                                                                                                            0.9s 
 => => writing image sha256:1085b060d72afe7f2b77b9f924c4f9269629b7fc686e01c0f45a51075fda9d84                                                       0.0s
 => => naming to docker.io/library/deps_test                                                                                                       0.0s

> docker run -it --entrypoint=/bin/sh deps_test
# ls
Dockerfile  data  dbt_packages	dbt_project.yml  logs  models  profiles.yml  target  tests  unused
# 
```

If it's possible can you provide your `packages` block from your project file so I can give a closer look?  It feels to me like `dbt deps` is failing somehow.

**Comment by alexrosenfeld10** [comment]
Thank you so much for taking a look @iknox-fa . What's really strange is it works for me on my local machine, but not in my CI job. However, my CI job runs the same commands... as per the log file I posted.

Here's my packages file, it's rather simple:
```yaml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.8.1
  - package: calogica/dbt_expectations
    version: 0.5.2
```

**Comment by alexrosenfeld10** [comment]
also, here's my `.dockerignore`:

```
pycharm_settings.png
README.md
Jenkinsfile
Dockerfile
.gitignore
.pre-commit-config.yaml
CODEOWNERS
ansible
scripts
```

**Comment by iknox-fa** [comment]
Lol those are two of the same ones I used to test.  Very strange.  I've reached the end of my day so I'll take another look with fresh eyes in the AM.  In the meantime I'm glad you have a workaround.

**Comment by alexrosenfeld10** [comment]
Yeah, I'm fried too I've been working on this all day haha. I hope we can get a consistent reproduction. Using the different directory seems to work, and I got my first successful deployed run of dbt with it so that's at least something! Thanks again for the help.

**Comment by alexrosenfeld10** [comment]
Btw, I tested this a tiny bit further - it definitely is because of the `VOLUME` statement. Any files created in a docker `VOLUME` as a result of a `RUN` command during the build process aren't allowed to persist:

```Dockerfile
FROM ghcr.io/dbt-labs/dbt-snowflake:1.0.latest

RUN dbt --version

RUN touch test.txt

WORKDIR /usr/my-project
COPY . .
RUN dbt deps

ENTRYPOINT ["dbt", "--log-format", "json", "--warn-error", "run"]
```

Then, playing around in the resulting container:

```
❯ docker run -it --entrypoint=/bin/sh <my-companies-jfrog-url>:2022-02-25.50-8d25a43-rel
# pwd
/usr/my-project
# ls
dbt_packages     logs    models	     seeds	tests
analyses      dbt_project.yml  macros  packages.yml  snapshots
# ls ..
app  bin  disc-dbt  games  include  lib  libexec  local  sbin  share  src
# cd ../app
# ls
dbt
# ls dbt
#
```

**Comment by alexrosenfeld10** [comment]
signing off for reals now 👋  catch'ya on this issue tomorrow or something

**Comment by leahwicz** [comment]
@alexrosenfeld10 thanks for reporting! At the moment we are not going to get time in the near future to dig into this more unfortunately. I'm going to add the `help_wanted` label on the issue in case anyone wants to take a crack at it before we free up. The [Docker image creation](https://github.com/dbt-labs/dbt-core/tree/main/docker) code is in the repo if people wanted to test out solutions or try something new.

**Comment by alexrosenfeld10** [comment]
@leahwicz ok, fair enough. I think it might be as simple as removing the `VOLUME /usr/app` line, but I don't have any tribal knowledge on why it's there in the first place. 🤷

**Comment by alexrosenfeld10** [comment]
at the very least, hopefully this issue serves as a searchable destination for users debugging the same weird behavior

**Comment by eugene-nikolaev** [comment]
> at the very least, hopefully this issue serves as a searchable destination for users debugging the same weird behavior

it does)

So is there any workaround?

**Comment by eugene-nikolaev** [comment]
Btw, it can change the volume state on Mac OS that's why it is even more confusing.

@jtcohen6 sorry for bugging, but what is the intended way to setup own dbt deps using a dbt docker image?

**Comment by alexrosenfeld10** [comment]
(whoops didn't mean to close)

**Comment by alexrosenfeld10** [comment]
No, I don't. I mean do this:

```Dockerfile
FROM ghcr.io/dbt-labs/dbt-snowflake:1.1.latest

# We have to move out of the default dir because of https://github.com/dbt-labs/dbt-core/issues/4784
WORKDIR /usr/wherever/you/want
COPY . .
RUN dbt deps

# warnings as errors - https://docs.getdbt.com/reference/global-configs#warnings-as-errors
# further arguments are passed in the kubernetes config as "args"
ENTRYPOINT ["dbt","--log-format", "json", "--warn-error", "run"]
```

**Comment by alexrosenfeld10** [comment]
ugh, the default action on command + enter is "comment and close". Obviously there are workarounds, but the core issue isn't fixed so i'm gonna reopen

**Comment by eugene-nikolaev** [comment]
@alexrosenfeld10, thanks a lot, works fine!

**Comment by yummydum** [comment]
Spent a whole day on the same problem, and finally I found this issue.
In my local Mac machine I can build with dbt_packages/ included in the image, but in the github action server, the dbt_packages/ are missing. 
At the minimum, this should be mentioned in the docs.

**Comment by indy-jonesy** [comment]
There is a different approach to this volume issue affecting `dbt_packages` and `dbt deps`.

I have built my own custom Dockerfile, but faced similar issues when trying to mount a `VOLUME` on the same directory of the active `dbt` project. Safe to say, `RUN dbt deps` was not working.

I was able to come up with a different approach to this problem that solved my use case.

You can simply update where `dbt deps` installs its packages, using the [packages-install-path](https://docs.getdbt.com/reference/project-configs/packages-install-path).

In the case of the current Dockerfile which has a `VOLUME` at `/usr/app`, simple pointing `packages-install-path` in the `dbt_project.yml` to a directory outside `/user/app` should be enough. I did it in my service accounts home directory, so `/home/dbt-user/dbt-packages`

TLDR:

1. We know the `VOLUME` is pointing to the `/user/app` directory
2. Update the `dbt_project.yml` to have a `packages-install-path: /home/dbt-user/dbt-packages` (or any path not associated with the `VOLUME`
3. Now, when you use `RUN dbt deps` it will package the installs in the appropriate directory and persist across the Docker layers.

**_Frankly, dbt labs should just update their image to place their packages outside of that directory to avoid this problem for everyone._**

**Comment by yehoshuadimarsky** [comment]
Running into the same issue, thanks everyone for your help. Just to add on to the previous solution, here's what I'm doing, using an env var (see the [jinja dbt docs](https://docs.getdbt.com/reference/dbt-jinja-functions/env_var)) to set the [packages-install-path](https://docs.getdbt.com/reference/project-configs/packages-install-path) and by default use the default one of `dbt_packages`. I only set this env var in the docker image, so I can control it fully.

In `dbt_project.yml`:
```yaml
clean-targets:
  - "target"
  - "{{ env_var('MYCOMPANY_DBT_PACKAGE_INSTALL_PATH', 'dbt_packages') }}"
packages-install-path: "{{ env_var('MYCOMPANY_DBT_PACKAGE_INSTALL_PATH', 'dbt_packages') }}"
```

My Dockerfile
```dockerfile
FROM ghcr.io/dbt-labs/dbt-bigquery:1.5.0

ENV MYCOMPANY_DBT_PACKAGE_INSTALL_PATH=/home/dbt_packages

RUN mkdir -m 777 -p ${MYCOMPANY_DBT_PACKAGE_INSTALL_PATH}

COPY ./my_package /usr/app
COPY ./profiles /root/.dbt/

RUN dbt deps
```

**Comment by alexrosenfeld10** [comment]
@jtcohen6 maybe it's time to revisit this issue? Seems like a lot of folks have this problem, and I don't think it'd be a major lift for dbt Labs to fix

**Comment by dbeatty10** [comment]
@alexrosenfeld10 would you be interested in raising a pull request, by any chance? e.g., a PR that removes this line (assuming that is the solution you are suggesting):

https://github.com/dbt-labs/dbt-core/blob/7740bd6b45b8a19cb5910fb1221bb7ed6d5e9e5d/docker/Dockerfile#L53

The other key thing we'd need to do is write up an explanation of what users can do to restore the previous behavior (if needed).
 
@sklirg may have already provided that [here](https://github.com/dbt-labs/dbt-core/issues/4784#issuecomment-1570221269):

> Furthermore, on the `VOLUME` train, there should be few affected by removing this directive. Specifying the `--volume` flag to a container runtime (e.g. `docker run --volume /opt/dbt:/usr/app`) is always allowed and will create a volume binding even though no VOLUME directive is set in the Dockerfile.

**Comment by alexrosenfeld10** [comment]
Sure, I can, but have little context into the impact / test cases / process needed.

**Comment by dbeatty10** [comment]
Here's a rough outline of the most important pieces when opening the PR:
1. Make your change in a local branch
1. [Install changie](https://github.com/dbt-labs/dbt-core/blob/a58b5ee8fb9fafd8bbf9da52d3b9e8603725592c/CONTRIBUTING.md#adding-or-modifying-a-changelog-entry), run `changie new`, and commit the file it creates
1. Sign the CLA
1. Open a PR with your change

You've already done most of those in https://github.com/dbt-labs/dbt-core/pull/8069 -- it's looking great 😎 

The main part that is missing is the changelog part -- I'll follow-up within the PR itself for next steps there.

Then a member of the dbt Labs engineering team will review the PR once it's open. They'll help figure out what kind of testing is needed and give any other feedback that is needed prior to merge.

**Comment by alexrosenfeld10** [comment]
yep, have done that before for actual code changes, just don't have the time on hand right now to get my local set up again (new machine). If you or someone else wants to push it over the line, that'd be fine, otherwise I'll get to it.. sometime

**Comment by dbeatty10** [comment]
No prob @alexrosenfeld10 -- I just added the changelog entry to the PR ✅

**Comment by alexrosenfeld10** [comment]
Thanks @dbeatty10
